File: adaptive_chat_server/app/status.py
"""Assembles the ``GET /status`` payload from the store and the live responder.

Deliberately carries no message text: the endpoint is unauthenticated and CORS
is wide open, so conversation content must not be reachable through it. What it
does report is volume (how many conversations, how many turns, how many tokens)
and the effective configuration the process is actually running.
"""
from __future__ import annotations

from app.stats import to_dict
from app.store import Conversation, ConversationStore


def _describe_responder(responder: object) -> dict:
    """Effective responder config, or a placeholder when it cannot report one.

    Tests swap in stubs via ``monkeypatch.setattr("app.main.responder", ...)``
    that need not implement the full protocol. A stub must not be able to turn
    /status into a 500, so a missing ``describe`` degrades instead of raising.
    """
    describe = getattr(responder, "describe", None)
    if not callable(describe):
        return {"kind": "unknown"}
    return describe()


def _conversation_row(conversation: Conversation) -> dict:
    """One conversation's volume: turn count, cumulative tokens, last turn."""
    prompt_tokens = 0
    reply_tokens = 0
    for iid in conversation.order:
        stats = conversation.interactions[iid].stats
        if stats is not None:
            prompt_tokens += stats.prompt_tokens
            reply_tokens += stats.reply_tokens

    last: dict | None = None
    if conversation.order:
        last_iid = conversation.order[-1]
        last_stats = conversation.interactions[last_iid].stats
        last = {
            "interactionId": last_iid,
            # None when that turn cost nothing measurable — echo mode or an
            # Ollama failure. The turn still counts toward interactionCount.
            "stats": to_dict(last_stats) if last_stats is not None else None,
        }

    return {
        "conversationId": conversation.conversation_id,
        "interactionCount": len(conversation.order),
        "totals": {
            "promptTokens": prompt_tokens,
            "replyTokens": reply_tokens,
            "totalTokens": prompt_tokens + reply_tokens,
        },
        "lastInteraction": last,
    }


def build_status(store: ConversationStore, responder: object) -> dict:
    """Operator snapshot of the running server.

    ``responder`` is typed loosely because the route passes the module-level
    singleton, which tests replace with stubs; only ``describe()`` is used.
    """
    conversations = store.list_conversations()
    return {
        "responder": _describe_responder(responder),
        "conversationCount": len(conversations),
        "conversations": [_conversation_row(c) for c in conversations],
    }

File: adaptive_chat_server_dart/lib/src/status.dart
/// Assembles the `GET /status` payload from the store and the live
/// responder.
///
/// Deliberately carries no message text: the endpoint is unauthenticated
/// and CORS is wide open, so conversation content must not be reachable
/// through it.
library;

import 'dart:convert';

import 'package:adaptive_chat_server_dart/src/responder.dart';
import 'package:adaptive_chat_server_dart/src/stats.dart';
import 'package:adaptive_chat_server_dart/src/store.dart';
import 'package:crypto/crypto.dart';

/// Stable, non-reversible label for correlating a conversation across polls.
///
/// The raw `conversationId` is the bearer credential for
/// `GET/POST /conversations/{cid}/interactions`, which return message text.
/// Since this endpoint is unauthenticated and CORS is wide open, leaking the
/// id here would hand any web page the developer visits a way to read the
/// transcript. Truncating the hash keeps it short while leaving it useless
/// for guessing the original id.
String conversationRef(String conversationId) {
  final digest = sha256.convert(utf8.encode(conversationId));
  return digest.toString().substring(0, 12);
}

/// Effective responder config, or a placeholder when it cannot report one.
///
/// A test double's `describe()` must not be able to turn `/status` into a
/// 500, so a throwing `describe()` degrades to `{'kind': 'unknown'}` instead
/// of propagating.
Map<String, dynamic> _describeResponder(Responder responder) {
  try {
    return responder.describe();
  } on Object {
    return {'kind': 'unknown'};
  }
}

Map<String, dynamic> _conversationRow(Conversation conversation) {
  var promptTokens = 0;
  var replyTokens = 0;
  for (final iid in conversation.order) {
    final stats = conversation.interactions[iid]!.stats;
    if (stats != null) {
      promptTokens += stats.promptTokens;
      replyTokens += stats.replyTokens;
    }
  }

  Map<String, dynamic>? last;
  if (conversation.order.isNotEmpty) {
    final lastIid = conversation.order.last;
    final lastStats = conversation.interactions[lastIid]!.stats;
    last = {'stats': lastStats != null ? statsToJson(lastStats) : null};
  }

  return {
    'conversationRef': conversationRef(conversation.conversationId),
    'interactionCount': conversation.order.length,
    'totals': {
      'promptTokens': promptTokens,
      'replyTokens': replyTokens,
      'totalTokens': promptTokens + replyTokens,
    },
    'lastInteraction': last,
  };
}

/// Operator snapshot of the running server.
Map<String, dynamic> buildStatus(
  ConversationStore store,
  Responder responder,
) {
  final conversations = store.listConversations();
  return {
    'responder': _describeResponder(responder),
    'conversationCount': conversations.length,
    'conversations': conversations.map(_conversationRow).toList(),
  };
}
